src/main/java/org/springframework/data/couchbase/core/convert/JsonValueConverter.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 the original author or authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.couchbase.core.convert;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.springframework.data.convert.PropertyValueConverter;
+import org.springframework.data.convert.ValueConversionContext;
+import org.springframework.data.mapping.PersistentProperty;
+
+/**
+ * Enum properties annotated with JsonValue
+ * 
+ * @author Michael Reiche
+ */
+public class JsonValueConverter
+		implements PropertyValueConverter<Object, Object, ValueConversionContext<? extends PersistentProperty<?>>> {
+
+	static protected Map<Class<?>, Method> valueMethodCache = new ConcurrentHashMap<>();
+	static protected Map<Class<?>, Constructor<?>> creatorMethodCache = new ConcurrentHashMap<>();
+	static protected Map<Class<?>, Map<Object, Enum<?>>> enumConstantsCache = new ConcurrentHashMap<>();
+
+	@Override
+	public Object read(Object value, ValueConversionContext<? extends PersistentProperty<?>> context) {
+		Class<?> type = context.getProperty().getActualType();
+
+		// if there was a @JsonCreator method, use it
+		if (getJsonCreatorMethod(type) != null) {
+			try {
+				return getJsonCreatorMethod(type).newInstance(value);
+			} catch (IllegalAccessException | InvocationTargetException | InstantiationException e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		// for Enums, we can use the inverse mapping we made from the @JsonValue method
+		if (Enum.class.isAssignableFrom(type)) {
+			return getEnumConstant((Class<Enum<?>>) type, value);
+		}
+
+		// Otherwise fall-through in MappingCouchbaseConverter.readValue()
+		throw new ConverterHasNoConversion();
```

Comment:
Filling in an exception's stack trace is relatively expensive. Could this potentially happen every time SDC deserializes an object?  Possible mitigation is to create a single instance of the exception, store it in a static final, and throw that instance over and over again.